tests/test_websockets.py

```diff
@@ -0,0 +1,151 @@
+
+    assert ev.is_set()
+    assert response.status == HTTPStatus.SWITCHING_PROTOCOLS
+
+    # With keep-alive in Connection
+    with liveclient as query:
+        response = await query('GET', '/ws', headers={
+            'Upgrade': 'websocket',
+            'Connection': 'keep-alive, upgrade',
+            'Sec-WebSocket-Key': 'hojIvDoHedBucveephosh8==',
+            'Sec-WebSocket-Version': '13'})
+
+    assert response.status == HTTPStatus.SWITCHING_PROTOCOLS
+
+
+@pytest.mark.asyncio
+async def test_websocket_communication(liveclient):
+
+    @liveclient.app.route('/echo', protocol="websocket")
+    async def echo(request, ws, **params):
+        while True:
+            message = await ws.recv()
+            await ws.send(message)
+
+    # Echo
+    websocket = await websockets.connect(liveclient.wsl + '/echo')
+    try:
+        for message in ('This', 'is', 'a', 'simple', 'test'):
+            await websocket.send(message)
+            echoed = await websocket.recv()
+            assert echoed == message
+    finally:
+        await websocket.close()
+
+    # Nothing sent, just close
+    websocket = await websockets.connect(liveclient.wsl + '/echo')
+    await websocket.close()
+
+
+@pytest.mark.asyncio
+async def test_websocket_broadcasting(liveclient):
+
+    @liveclient.app.route('/broadcast', protocol="websocket")
+    async def feed(request, ws, **params):
+        while True:
+            message = await ws.recv()
+            await asyncio.wait([
+                socket.send(message) for socket
+                in request.app.websockets if socket != ws])
+
+    # connecting
+    connected = []
+    for n in range(1, 6):
+        ws = await websockets.connect(liveclient.wsl + '/broadcast')
+        connected.append(ws)
+
+    # Broadcasting
+    for wid, ws in enumerate(connected, 1):
+        broadcast = 'this is a broadcast from {}'.format(wid)
+        await ws.send(broadcast)
+        for ows in connected:
+            if ows != ws:
+                message = await ows.recv()
+                assert message == broadcast
+
+    # Closing
+    for ws in connected:
+        await ws.close()
+
+
+@pytest.mark.asyncio
+async def test_websocket_binary(liveclient):
+
+    @liveclient.app.route('/bin', protocol="websocket")
+    async def binary(request, ws, **params):
+        await ws.send(b'test')
+
+    # Echo
+    websocket = await websockets.connect(liveclient.wsl + '/bin')
+    bdata = await websocket.recv()
+    await websocket.close_connection_task
+    assert bdata == b'test'
+    assert websocket.close_reason == ''
+    assert websocket.state == websockets.protocol.State.CLOSED
+
+
+@pytest.mark.asyncio
+async def test_websocket_route_with_subprotocols(liveclient):
+    results = []
+
+    @liveclient.app.route(
+        '/ws', protocol="websocket", subprotocols=['foo', 'bar'])
+    async def handler(request, ws):
+        results.append(ws.subprotocol)
+
+    with liveclient as query:
```

Comment:
Shower thought: I wonder if we can close the client from the fixture instead of needing a context manager for each query.

Reply:
If I recall correctly our discussion, I think we decided to align the `liveclient` to the `client` one, so instead of the context manager, simply do something like:

    await liveclient.query(…)